### PR TITLE
[SYCL] Enable some math builtins in SemaSYCL

### PR DESCRIPTION
**Pull request by jinge90**
Signed-off-by: gejin <ge.jin@intel.com>

This PR aims to enable some math builtin in SemaSYCL. We supported some llvm math intrinsic in this PR:
https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/855
So a bunch of math builtin which depend on those llvm math intrinsic can work now. This PR enables following math builtin:
fmax/f (depends on llvm.maxnum)
fmin/f  (depends on llvm.minnum)
isinf     (depends on llvm.fabs)
isfinite  (depends on llvm.fabs)
isnormal (depends on llvm.fabs)
fpclassify (depends on llvm.fabs)

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -178,20 +178,8 @@ static bool IsSyclMathFunc(unsigned BuiltinID) {
...
-  case Builtin::BI__builtin_fmax:
-  case Builtin::BIfmin:
-  case Builtin::BI__builtin_fmin:
-  case Builtin::BIfmaxf:
-  case Builtin::BI__builtin_fmaxf:
-  case Builtin::BIfminf:
-  case Builtin::BI__builtin_fminf:
   case Builtin::BIlroundf:
   case Builtin::BI__builtin_lroundf:
-  case Builtin::BI__builtin_fpclassify:
```
Comment (bader) [inline clang/lib/Sema/SemaSYCL.cpp:191]: This built-in is not tested.

Comment (jinge90) [reply clang/lib/Sema/SemaSYCL.cpp:191]: Hi. @bader 
The test for fpclassify builtin is added.
Thanks.

File: clang/test/SemaSYCL/supported_math.cpp [hunk 2]
```diff
@@ -26,39 +30,51 @@ int main() {
...
-    acc[0] += (int)__builtin_rintf(1.0f);            // expected-no-diagnostics
-    acc[0] += (int)__builtin_rint(1.0);              // expected-no-diagnostics
-    acc[0] += (int)__builtin_nearbyintf(0.5f);       // expected-no-diagnostics
-    acc[0] += (int)__builtin_nearbyint(0.5);         // expected-no-diagnostics
-    acc[0] += (int)__builtin_floorf(0.5f);           // expected-no-diagnostics
-    acc[0] += (int)__builtin_floor(0.5);             // expected-no-diagnostics
-    acc[0] += (int)__builtin_copysignf(1.0f, -0.5f); // expected-no-diagnostics
-    acc[0] += (int)__builtin_logf(1.0f);             // expected-no-diagnostics
-    acc[0] += (int)__builtin_log(1.0);               // expected-no-diagnostics
+    acc[0] += (int)truncf(1.0f);                        // expected-no-diagnostics
```
Comment (bader) [inline clang/test/SemaSYCL/supported_math.cpp:33]: expected-no-diagnostics can be placed only once
Here is an example: https://github.com/llvm/llvm-project/blob/main/clang/test/Sema/aarch64-sve-explicit-casts-fixed-size.c

Comment (jinge90) [reply clang/test/SemaSYCL/supported_math.cpp]: Hi, @bader 
Thanks for the reminder, just fixed it.

**Review by bader — APPROVED** [review]
LGTM, although I would change the title of the PR.
"[SYCL] Enable some math builtins in SemaSYCL" - most of the enabled built-ins are not from "math" category.

**Review by elizabethandrews — APPROVED** [review]

**Comment by keryell** [comment]
@Ralender would this solve some crashes we have on math stuff?

**Comment by jinge90** [comment]
> @Ralender would this solve some crashes we have on math stuff?

Hi, @keryell 
I think this PR will not fix any "crash" issue. Previously, those math builtin are forbidden in FE, if developers use them in SYCL device code, compiler will report compiling error in device compilation phase.  Now, those math builtin are permitted which means device compiler will convert them to corresponding llvm intrinsic and no compiling error will be reported.
The reason why we disabled some math builtin is they will be converted to corresponding llvm intrinsic which can't be handled by llvm-spirv translator. Handling such llvm intrinsic will lead to llvm-spirv crash.
Before submitting this patch, we have submitted a patch to enable a bunch of llvm intrinsic in llvm-spirv which can solve some "crash" in llvm-spirv side: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/855
Thanks very much.

**Comment by keryell** [comment]
I was not thinking to Intel related crashes but to Xilinx related crashes.
We are not using SPIR-V but another path which has some creative/unexpected/... effects on the Xilinx FPGA backend.

**Pull request merged**